Epoch data channel

lstipakov opened the pull request.

lstipakov committed:
peer: add spin lock helpers

Introduce OvpnAcquire/ReleaseSpinLock(), which
acquire/release spinlock taking into account dpc
and shared/exclusive.

Signed-off-by: Lev Stipakov <lev@openvpn.net>

lstipakov committed:
crypto: add epoch data channel

This implements epoch data channel format for
AEAD ciphers, based on RFC and userspace implementation.

 - https://openvpn.github.io/openvpn-rfc/openvpn-wire-protocol.html#name-openvpn-epoch-data-channel-
 - https://gerrit.openvpn.net/c/openvpn/+/806

The differences with userspace implemenation are:

 - use BCrypt key derivation API instead of own implementation
 - receive key usage limits are ignored

Note that this lacks proper locking on updating epoch keys,
this will be implemented in follow-up commit.

Driver version is bumped to 2.8.0.

GitHub: https://github.com/OpenVPN/ovpn-dco-win/issues/104

Signed-off-by: Lev Stipakov <lev@openvpn.net>

lstipakov committed:
crypto: take exclusive lock for epoch transitions

AEAD needs to tweak its epoch keys whenever we hit the usage limit (TX)
or see a packet from a new epoch (RX). That update isn’t
atomic: we rebuild send/recv contexts, refresh future keys,
and advance nonce counters. The code used to do this while the
peer spin lock was held in shared mode, so another thread
could see half-written state.

Introduce OvpnCryptoCallWithRetry() plus small invoke helpers.
The crypto routines keep working under the shared lock on the fast path,
but they return STATUS_OVPN_CRYPTO_RETRY when a real mutation is required.
The caller then upgrades to the exclusive lock, retries with allowRekey = TRUE,
and commits the changes once they’re consistent. That keeps normal traffic
on the cheap shared lock while epoch rotations happen safely under exclusive ownership.

Signed-off-by: Lev Stipakov <lev@openvpn.net>

lstipakov commented:
@mfil it would be nice if you could check (at least) the crypto part. It mirrors the userspace implementation.

mfil commented on crypto.cpp:
```diff
@@ -123,126 +260,498 @@ OvpnCryptoUninitAlgHandles(_In_ BCRYPT_ALG_HANDLE aesAlgHandle, BCRYPT_ALG_HANDL
+
+    static const char prefix[] = "ovpn ";
+
+    size_t prefixLen = 0;
```
Why not make it `const size_t prefixLen = sizeof(prefix) - 1`?

mfil commented on crypto.cpp:
```diff
@@ -123,126 +260,498 @@ OvpnCryptoUninitAlgHandles(_In_ BCRYPT_ALG_HANDLE aesAlgHandle, BCRYPT_ALG_HANDL
+    GOTO_IF_NOT_NT_SUCCESS(done, status, BCryptKeyDerivation(hKey, &desc, outBytes, outLen, &got, 0));
+
+    if (got != outLen) {
+        status = STATUS_BUFFER_TOO_SMALL;
```
I'm not sure when this error would actually happen. HKDF will try to put `outLen` bytes into `outBytes`, no matter how big or small that is. (Within reason. I think the maximum is 255 * 32 bytes.) If the buffer is actually shorter than `outLen`, it's undefined behavior.

Basically, what I'm saying is, there's no such thing as a too small HKDF buffer. If you ask it for 3 bytes of key material, that's what you get. It won't say 3 is too small. So I don't know what would be an appropriate error code here. Seems like it would have to be some internal error in BCryptKeyDerivation. Or maybe if the caller asks for a huge amount of key material.

lstipakov replied:
I see. This parameter is passed from the userspace and driver only ensures that it is <= 32. With the smaller and possibly incorrect value we should not crash but key generation fails, which we'll catch. I agree with you' will remove this check.

mfil replied:
I'm not saying the check should be removed (I don't know), just that *if* this happens, it's not because the buffer is too small.

lstipakov committed:
crypto: fix cipher "none" new key setup

Uninit key slot, set keId/peerId and reset crypto options.

Signed-off-by: Lev Stipakov <lev@openvpn.net>

mfil commented on crypto.cpp:
```diff
@@ -124,125 +251,491 @@ OvpnCryptoUninitAlgHandles(_In_ BCRYPT_ALG_HANDLE aesAlgHandle, BCRYPT_ALG_HANDL
+
+    // derive
+    ULONG got = 0;
+    GOTO_IF_NOT_NT_SUCCESS(done, status, BCryptKeyDerivation(hKey, &desc, outBytes, outLen, &got, 0));
```
I'm not sure if removing the `outLen != got` check was the right move. I was disagreeing with the error code you returned in that situation.

If there's a situation where this function generates less than `outLen` bytes, we absolutely need to catch that! Given the way HKDF is defined, I just don't see a valid reason for the function to do that. So I don't know what an appropriate error code would be.

lstipakov replied:
I see, I don't expect it to happen, but I guess having this check won't harm. Added it back.

mfil commented on crypto.cpp:
```diff
@@ -123,126 +250,495 @@ OvpnCryptoUninitAlgHandles(_In_ BCRYPT_ALG_HANDLE aesAlgHandle, BCRYPT_ALG_HANDL
+    ULONG got = 0;
+    GOTO_IF_NOT_NT_SUCCESS(done, status, BCryptKeyDerivation(hKey, &desc, outBytes, outLen, &got, 0));
+    if (got != outLen) {
+        status = STATUS_BUFFER_TOO_SMALL;
```
Can we make this `STATUS_INTERNAL_ERROR` or something?

lstipakov replied:
Will do,

schwabe commented on pktid.cpp:
```diff
@@ -28,23 +28,17 @@
 
 _Use_decl_annotations_
-NTSTATUS OvpnPktidXmitNext(OvpnPktidXmit* px, VOID* pktId, BOOLEAN pktId64bit)
+NTSTATUS OvpnPktidXmitNext(OvpnPktidXmit* px, VOID* pktId)
```
Maybe add here a comment taht this is only for the non-epoch use case?

lstipakov replied:
Will do.

schwabe commented on crypto.cpp:
```diff
@@ -123,126 +133,484 @@ OvpnCryptoUninitAlgHandles(_In_ BCRYPT_ALG_HANDLE aesAlgHandle, BCRYPT_ALG_HANDL
+            }
+
+            // calculate 64-bit packet-id = (epoch << 48) | ctr48
+            UINT64 ctr48 = InterlockedIncrementNoFence64(&keySlot->PktidXmit.SeqNum) & 0x0000FFFFFFFFFFFFULL;
```
There is no overflow checky here, you are just masking the bits. This need to either checked here or a comment to say where we ensure that this never overflows.

lstipakov replied:
It is checked above:

`if (OvpnCryptoAeadUsageLimitReached(opts->AeadUsageLimit, keySlot->Encrypt.PlaintextBlocks, keySlot->PktidXmit.SeqNum) || (keySlot->PktidXmit.SeqNum == PACKET_ID_EPOCH_MAX)) {
`

I'll mention it in the comment.

schwabe commented on crypto.cpp:
```diff
@@ -123,126 +133,484 @@ OvpnCryptoUninitAlgHandles(_In_ BCRYPT_ALG_HANDLE aesAlgHandle, BCRYPT_ALG_HANDL
+VOID
+OvpnCryptoMakeEpochNonce(UCHAR *epochIv, UINT64 packet_id_net, UCHAR *nonce)
+{
+    // first 8 bytes of IV (aka nonce) is pktid_net XOR 8 bytes of epoch IV
```
8 bytes would be 64bit. But Epoch IV is only 48bits

lstipakov replied:
Hm.. no? IV is 12 bytes, you write yourself in RFC: 

`The 96bit IV is constructed by using the 96 bit of implicit IV belonging to the epoch and doing an XOR with the 64 bit packet id of both epoch and epoch_counter`

schwabe commented on crypto.cpp:
```diff
@@ -123,126 +133,484 @@ OvpnCryptoUninitAlgHandles(_In_ BCRYPT_ALG_HANDLE aesAlgHandle, BCRYPT_ALG_HANDL
+    iv0 ^= packet_id_net;
+    RtlCopyMemory(nonce, &iv0, sizeof(iv0));
+
+    // last 4 bytes of IV are from epoch IV
```
Do you mean the epoch itself?

lstipakov replied:
Epoch's implicit IV. As RFC says, `implicit IV belonging to the epoch`. Clarified in comments.

schwabe commented on crypto.cpp:
```diff
@@ -123,126 +133,484 @@ OvpnCryptoUninitAlgHandles(_In_ BCRYPT_ALG_HANDLE aesAlgHandle, BCRYPT_ALG_HANDL
+// } OvpnLabel;
+_IRQL_requires_max_(PASSIVE_LEVEL)
+static
+NTSTATUS OvpnCryptoMakeLabel(
```
This is probably a good candidcate for a unit test.

lstipakov replied:
Yes, will add unit tests next, preferably in a separate PR, as it requires some changes to the project structure.

lstipakov replied:
Done!

schwabe commented on crypto.cpp:
```diff
@@ -123,126 +133,484 @@ OvpnCryptoUninitAlgHandles(_In_ BCRYPT_ALG_HANDLE aesAlgHandle, BCRYPT_ALG_HANDL
+    UINT16 rx_epoch = 0;
+
     UCHAR nonce[12];
     if (encrypt) {
```
Maybe split this method into encrypt and decrypt? This has become quite large by now.

lstipakov replied:
Yes, will do. In fact there are already separate Encrypt and Decrypt functions, which both call DoWork. I agree that the single function now makes less sense.

schwabe commented on crypto.h:
```diff
@@ -115,3 +162,30 @@ UCHAR OvpnCryptoOpcodeExtract(UCHAR op)
+        return 0;
+
+    default:
+        return (1ull << 36) - 1;
```
Maybe add a comment here that this is the limit for AES-GCM

lstipakov replied:
Done!

lstipakov committed:
crypto: split encrypt and decrypt code

The common function OvpnCryptoAEADDoWork() has grown too large
and now it doesn't make sense to have it. Do encrypt/decrypt
in itn own functions.

While on it, improve comments.

Signed-off-by: Lev Stipakov <lev@openvpn.net>

lstipakov committed:
Add unit tests for epoch crypto logic

Create a separate googletest-based project.
Move some of epoch crypto code into own compilation unit.
Add test run to GHA.

Signed-off-by: Lev Stipakov <lev@openvpn.net>

mfil commented:
The crypto code looks good to me, but I don't feel like I understand the Windows driver stuff enough to approve. Can we get someone else to take another look?
